roofit/roofitcore/src/RooAbsCollection.cxx

```diff
@@ -1281,3 +1292,67 @@ std::unique_ptr<RooAbsCollection::LegacyIterator_t> RooAbsCollection::makeLegacy
   return std::make_unique<LegacyIterator_t>(_list);
 }
 
+
+////////////////////////////////////////////////////////////////////////////////
+/// Insert an element into the owned collections.
+void RooAbsCollection::insert(RooAbsArg* item) {
+  _list.push_back(item);
+
+  if (_allRRV && dynamic_cast<const RooRealVar*>(item)==0) {
+    _allRRV= false;
+  }
+
+  if (_nameToItemMap) {
+    (*_nameToItemMap)[item->namePtr()] = item;
+  }
+}
+
+
+////////////////////////////////////////////////////////////////////////////////
+/// Install an internal hash map for fast finding of elements by name.
+void RooAbsCollection::useHashMapForFind(bool flag) const {
+  if (!flag && _nameToItemMap){
+    _nameToItemMap = nullptr;
+  }
+
+  if (flag && !_nameToItemMap) {
+    _nameToItemMap.reset(new std::unordered_map<const TNamed*, Storage_t::value_type>());
+    for (const auto item : _list) {
+      (*_nameToItemMap)[item->namePtr()] = item;
+    }
+  }
+}
+
+
+////////////////////////////////////////////////////////////////////////////////
+/// Perform a search in a hash map.
+/// This only happens if this collection is larger than _sizeThresholdForMapSearch.
+/// This search is *not guaranteed* to find an existing
+/// element because elements can be renamed while
+/// being stored in the collection.
+RooAbsArg* RooAbsCollection::tryFastFind(const TNamed* namePtr) const {
+  if (_list.size() >= _sizeThresholdForMapSearch && !_nameToItemMap) {
+    useHashMapForFind(true);
+    assert(_nameToItemMap);
+  }
+
+  if (!_nameToItemMap) {
+    return nullptr;
+  }
+
+  auto item = _nameToItemMap->find(namePtr);
+  if (item != _nameToItemMap->end()) {
+    // Have an element. Check that it didn't get renamed.
+    if (item->second->namePtr() == item->first) {
+      return item->second;
+    } else {
+      // Item has been renamed / replaced.
+      _nameToItemMap->erase(item);
+      if (auto arg = findUsingNamePointer(_list, namePtr)) {
+        (*_nameToItemMap)[arg->namePtr()] = arg;
+      }
```

Comment:
so the output of `tryFastFind` can be that `_nameToItemMap` is updated to have the element with `namePtr`, but the function returns nullptr anyway..? i.e. I would expect a `return arg` here

Reply:
Good catch, thanks!
It will still work, but it's now doing a linear search twice.